i6_models/parts/frontend/vgg_act.py

```diff
@@ -0,0 +1,197 @@
+            model_cfg.conv_padding if model_cfg.conv_padding is not None else _get_padding(model_cfg.conv_kernel_size)
+        )
+        pool1_padding = model_cfg.pool1_padding if model_cfg.pool1_padding is not None else 0
+        pool2_padding = model_cfg.pool2_padding if model_cfg.pool2_padding is not None else 0
+
+        self.include_linear_layer = True if model_cfg.linear_output_dim is not None else False
+        assert not (
+            (model_cfg.linear_input_dim is not None) != (model_cfg.linear_output_dim is not None)
+        ), "please set input and output dim of the linear layer"
+
+        self.conv1 = nn.Conv2d(
+            in_channels=1,
+            out_channels=model_cfg.conv1_channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.conv2 = nn.Conv2d(
+            in_channels=model_cfg.conv1_channels,
+            out_channels=model_cfg.conv2_channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.pool1 = nn.MaxPool2d(
+            kernel_size=model_cfg.pool1_kernel_size,
+            stride=model_cfg.pool1_stride,
+            padding=pool1_padding,
+        )
+        self.conv3 = nn.Conv2d(
+            in_channels=model_cfg.conv2_channels,
+            out_channels=model_cfg.conv3_channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.conv4 = nn.Conv2d(
+            in_channels=model_cfg.conv3_channels,
+            out_channels=model_cfg.conv4_channels,
+            kernel_size=model_cfg.conv_kernel_size,
+            padding=conv_padding,
+        )
+        self.pool2 = nn.MaxPool2d(
+            kernel_size=model_cfg.pool2_kernel_size,
+            stride=model_cfg.pool2_stride,
+            padding=pool2_padding,
+        )
+        self.activation = model_cfg.activation
+        if self.include_linear_layer:
+            self.linear = nn.Linear(
+                in_features=model_cfg.linear_input_dim,
+                out_features=model_cfg.linear_output_dim,
+                bias=True,
+            )
+
+    def forward(self, tensor: torch.Tensor, sequence_mask: Optional[torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        T might be reduced to T' or T'' depending on stride of the layers
+
+        :param tensor: input tensor of shape [B,T,F]
+        :param sequence_mask: the sequence mask for the tensor
+        :return: torch.Tensor of shape [B,T",F'] and the shape of the sequence mask
+        """
+        # and add a dim
+        tensor = tensor[:, None, :, :]  # [B,C=1,T,F]
+
+        tensor = self.conv1(tensor)
+        tensor = self.conv2(tensor)
+
+        tensor = self.activation(tensor)
+        tensor = self.pool1(tensor)  # [B,C,T',F']
+        if sequence_mask is not None:
+            sequence_mask = sequence_mask.float()
+            sequence_mask = _mask_pool(
```

Comment:
Before you do the mask pooling for `pool1`, you need to do the mask pooling for `conv1` and `conv2`. See also my earlier (yet unresolved) comments on this. There are also more mask poolings missing here.

Reply:
the stride controls the mask size if I understand correctly.. so I should only update the mask when I allow setting the stride. since I am following our implementation I do not allow this everywhere.

Reply:
If the stride is one and same padding, the `_mask_pool` does anyway not have an effect. I would prefer writing code in a way that reduces the amount if implicit assumptions on other parts. Write it in a way that it will just always work. Or at least make a comment here explaining exactly what you say, so that a reader does not need to know it. When reading some code, you should be able to understand the code just by looking at it, without knowing the surrounding code, so the logic should alway be as local as possible, with the least amount of dependencies and assumptions, or any implicit assumptions should be very clearly mentioned, otherwise the code looks wrong.

Reply:
`_mask_pool(..., self.conv1.stride, ...)` will just always work, no matter what striding (or no striding) you have in `self.conv1`.

Reply:
just for my understanding :)
from the code, as is there should be no striding.. but you could set it via `model.conv1.stride` from the outside or so from the outside, correct? that would cover that edge case, correct? are there any other edge cases that you are thinking of?

Reply:
I'm not speaking at all about any such edge cases. I mean for a reader of the code, who does not know anything about how `conv1.stride` is configured, the code looks wrong. So to understand this code, you need to have such external knowledge. This is usually a bad idea. You want to reduce interdependencies, to be able to understand code locally, without needing to know anything else. It just makes code easier to understand, and also less buggy overall. E.g. now assume you change the other code, or maybe you copy over this module and take it as a base for a V2 variant, where the `conv1.stride` is configurable and not fixed. You can easily forget then about that you also need to change this other code, which depends on this specific assumption. It's always good to remove interdependencies.

Reply:
I see your point.
follow up: what is external knowledge for you? what is locally? everything not in the function? because the config and module should always go together or not?!

Reply:
Everything in the function. And there, the closer in number-of-lines-difference, the better. It's a continuum. Going to the other extreme, yes, in the same file is still better than across files. You cannot always avoid it. There always need to be compromises. Sometimes it would have an effect on efficiency and you don't want that.

Btw, here, for efficiency, in your `_mask_pool` function, you might want to add an extra check if `stride == 1` and same padding that then you don't need to do anything.